File: apple/YAOpenEvidence/Features/Answer/JobLiveMonitor.swift
import Foundation
import YAOEKit

/// 任务事件流的连接与重连管理。退避 1s 起 ×2，上限 10s，不因成功连接而重置。
@MainActor
@Observable
final class JobLiveMonitor {
    enum Connection: Equatable {
        case idle
        case connecting
        case open
        case reconnecting
        case closed
    }

    private(set) var live = JobLive.empty
    private(set) var connection: Connection = .idle

    /// 是否已在订阅中：页面重新出现时用它避免重复 `start()`。
    var isRunning: Bool { task != nil }

    private let jobID: String
    private let client: APIClient
    private let onEvent: (JobLive, SSEEvent) -> Void
    private var task: Task<Void, Never>?
    /// 已消费到的 Redis Stream 位置：暂停后再续订从这里继续，避免重放造成日志重复。
    private var lastEventID = "0-0"
    /// 暂停后再续订算新一代：旧任务收尾时不得把新任务的 `task` 清空。
    private var generation = 0

    init(
        jobID: String,
        client: APIClient,
        onEvent: @escaping (JobLive, SSEEvent) -> Void
    ) {
        self.jobID = jobID
        self.client = client
        self.onEvent = onEvent
    }

    /// 订阅事件流。已终态或已在运行时是空操作，可在页面重新出现时安全重复调用。
    func start() {
        guard task == nil, live.terminal == nil else { return }
        generation += 1
        let generation = generation
        task = Task { [weak self] in
            await self?.run()
            guard let self, self.generation == generation else { return }
            task = nil
        }
    }

    /// 暂停订阅（离开页面）。保留 `live` 与 `lastEventID`，回到页面后 `start()` 可续订。
    func stop() {
        task?.cancel()
        task = nil
        connection = live.terminal == nil ? .idle : .closed
    }

    private func run() async {
        var delay = Duration.seconds(1)
        var first = true

        while !Task.isCancelled {
            connection = first ? .connecting : .reconnecting
            do {
                for try await event in await client.events(jobID: jobID, lastEventID: lastEventID) {
                    if let id = event.id { lastEventID = id }
                    connection = .open
                    live = live.applying(event)
                    onEvent(live, event)
                }
            } catch {
                if Task.isCancelled { return }
                // 鉴权/参数类错误重连也不会好转。
                if let status = (error as? APIError)?.status, [401, 403, 404, 422].contains(status) {
                    connection = .closed
                    return
                }
            }

            if live.terminal != nil {
                connection = .closed
                return
            }
            if Task.isCancelled { return }

            first = false
            connection = .reconnecting
            // 重连前探活：令牌已失效就停连，不再空转重试。会话由 `APIClient`
            // 的 onUnauthorized 统一清理（那里带会话代次，不会误杀新登录）。
            do {
                _ = try await client.me()
            } catch {
                if error.status == 401 {
                    connection = .closed
                    return
                }
            }

            try? await Task.sleep(for: delay)
            delay = min(delay * 2, .seconds(10))
        }
    }
}

File: flutter/lib/features/answer/job_live_monitor.dart
import 'dart:async';
import 'dart:math' as math;

import 'package:freezed_annotation/freezed_annotation.dart';
import 'package:riverpod_annotation/riverpod_annotation.dart';

import '../../core/api/api_error.dart';
import '../../core/api/endpoints.dart';
import '../../core/api/sse.dart';
import '../../core/logic/job_live.dart';
import '../../core/session/session_controller.dart';

part 'job_live_monitor.freezed.dart';
part 'job_live_monitor.g.dart';

/// SSE 连接状态（答案页指示灯）。
enum SseConnection {
  idle,
  connecting,
  open,
  reconnecting,
  closed;

  /// 只有「在等」和「出问题了」值得占一行字；连接正常是默认预期，说出来是噪音。
  String? get label => switch (this) {
    SseConnection.connecting => '连接中',
    SseConnection.reconnecting => '重连中',
    SseConnection.idle || SseConnection.open || SseConnection.closed => null,
  };
}

@freezed
abstract class JobLiveState with _$JobLiveState {
  const factory JobLiveState({
    @Default(JobLive.empty) JobLive live,
    @Default(SseConnection.idle) SseConnection connection,
    @Default('0-0') String lastEventId,
  }) = _JobLiveState;
}

/// 单个任务的事件流监听器：断线重连 + 事件归约。
///
/// 重连策略与 Apple 端 `JobLiveMonitor` 一致：退避 1→2→4→8→10 s 封顶；
/// 重连前先 `me()` 探活，401 直接收摊（会话已失效，重试只会刷 401）。
@riverpod
class JobLiveMonitor extends _$JobLiveMonitor {
  static const _fatalStatuses = {401, 403, 404, 422};

  bool _cancelled = false;
  StreamSubscription<SseEvent>? _subscription;
  Completer<void>? _gate;

  @override
  JobLiveState build(String jobId) {
    ref.onDispose(_stop);
    Future(() => _run(jobId));
    return const JobLiveState();
  }

  void _stop() {
    _cancelled = true;
    _subscription?.cancel();
    _subscription = null;
    if (_gate?.isCompleted == false) _gate!.complete();
  }

  Future<void> _run(String jobId) async {
    final client = ref.read(apiClientProvider);
    var delay = const Duration(seconds: 1);

    while (!_cancelled) {
      state = state.copyWith(connection: SseConnection.connecting);
      var fatal = false;
      final gate = Completer<void>();
      _gate = gate;
      try {
        final stream = client.events(
          jobId: jobId,
          lastEventId: state.lastEventId,
          onOpen: () {
            if (!_cancelled) {
              state = state.copyWith(connection: SseConnection.open);
            }
          },
        );
        _subscription = stream.listen(
          _onEvent,
          onError: gate.completeError,
          onDone: () {
            if (!gate.isCompleted) gate.complete();
          },
          cancelOnError: true,
        );
        await gate.future;
      } on ApiError catch (error) {
        fatal = _fatalStatuses.contains(error.status);
      } catch (_) {
        // 传输层异常：按可重连处理。
      } finally {
        await _subscription?.cancel();
        _subscription = null;
        _gate = null;
      }

      if (_cancelled) return;
      // 终态已到达或错误不可恢复：收摊。
      if (fatal || state.live.terminal != null) {
        state = state.copyWith(connection: SseConnection.closed);
        return;
      }

      state = state.copyWith(connection: SseConnection.reconnecting);
      try {
        await client.me();
      } on ApiError catch (error) {
        if (error.status == 401) {
          state = state.copyWith(connection: SseConnection.closed);
          return;
        }
      }
      if (_cancelled) return;
      await Future<void>.delayed(delay);
      delay = Duration(
        milliseconds: math.min(delay.inMilliseconds * 2, 10000),
      );
    }
  }

  void _onEvent(SseEvent event) {
    if (_cancelled) return;
    final live = state.live.applying(event);
    state = state.copyWith(
      live: live,
      lastEventId: event.id ?? state.lastEventId,
    );
    if (live.terminal != null) {
      // 终态帧到达后主动断流，不等服务端关闭。
      _subscription?.cancel();
      _subscription = null;
      if (_gate?.isCompleted == false) _gate!.complete();
    }
  }
}
